dist/tools/mkconstfs/mkconstfs2.py

```diff
@@ -0,0 +1,160 @@
+    .private_data = (void *)&_fs_data,
+}};
+"""
+
+FILES_DECL = """
+static const constfs_file_t _files[] = {
+"""
+
+BLOB_DECL = """
+/** {fname} **/
+static const uint8_t {varname}[] = {{
+"""
+
+
+def mkconstfs(files, root_path, mount_point, constfs_name):
+    """Generate a C file containing a constant file system
+
+    Return
+    ------
+
+    chunks: Iterator yielding fragments of the of the output file.
+    """
+
+    filemap = {f: os.path.relpath(f, root_path) for f in files}
+
+    yield C_HEADER
+    yield from itertools.chain.from_iterable(
+                print_file_data(*fl) for fl in filemap.items())
+
+    yield FILES_DECL
+
+    yield from (FILE_TEMPLATE.format(target_name=relp, mangled_name=mangle_name(relp))
+                for relp in sorted(filemap.values()))
+
+    yield "};\n"
+
+    yield C_FOOTER.format(constfs_name=constfs_name, mount_pount=mount_point)
+
+
+def mangle_name(fname):
+    return "_"+"".join(c if c.isalnum() else "_{:04x}".format(ord(c))
+                       for c in fname)
+
+
+def print_file_data(local_fname, target_fname):
+    """Convert a file into a static C array:
+
+    Parameters
+    ----------
+
+    local_fname: real file name (where the file is on this machine's fs)
+    target_fname: name that the file will have in the constfs.
+    output_file: File-like object where the array will be written.
+
+    Return
+    ------
+
+    chunks: Iterator yielding fragments of the of the output text.
+    """
+
+    varname = mangle_name(target_fname)
+    yield BLOB_DECL.format(fname=target_fname, varname=varname)
+
+    def byte2s(b):
+        return "0x{},".format(hexlify(b).decode('utf-8'))
+
+    with open(local_fname, 'rb') as f, \
+         mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ) as bfile:  # noqa: E127
+        yield from map(lambda x: x[1],
+            itertools.chain.from_iterable(     # noqa: E128
+                map(lambda l: itertools.chain(l[1], [(0, "\n")]),
+                    itertools.groupby(
+                        enumerate(map(byte2s, bfile)),
+                        lambda x: x[0]//16)
+                )  # noqa: E127
+            )
+        )
+
+    yield "};\n"
+
+
+def main():
+    parser = argparse.ArgumentParser(
+            description="Embed files into a constant file system")
+
+    parser.add_argument("-m", '--mount', metavar="mountpoint",
+                        help="Where to mount the resulting fs", default="/")
+
+    parser.add_argument("-o", '--output', metavar="output_file",
+                        help="Write the output to a file instead of stdout."
+                             "The file is only written if the command is successful"
+                             "(i.e. there is no partial output")
+
+    parser.add_argument("-r", '--root', metavar="root_base_path",
+                        help="Paths on the constf will be gerated for the real"
+                        "path of the files by considering this path to be the root"
```

Comment:
insert whitespaces at the end of the multiline strings above (i.e. after _root_, _real_, ..)